### ISSUE TITLE
Docker container restarts randomly

### ISSUE THREAD

**Issue by dabide** [DarthSim/imgproxy#60]
We observe random restarts when running the Docker container:

```
2018/05/28 10:09:01 [ZCn...Bn.jpg
2018/05/28 10:09:02 Starting server at :8080
2018/05/28 10:09:03 [w3m...Bn.jpg
2018/05/28 10:09:03 [Had...
```

This happens at least a couple of times per hour.

**Comment by DarthSim** [comment]
Hi David,

I'm sure this depends on container system you use. Afaik Heroku does it if you use it for free. Our imgproxy containers live for months without any restarts.

**Comment by dabide** [comment]
Hi, @DarthSim,

We are using Docker CE on Ubuntu.

**Comment by DarthSim** [comment]
I'm still sure there is a problem with your container system. Еhe official docker image was downloaded for thousands of times and you're the first who reported the issue. Try to take a look at the docker logs, maybe there are some clues.

**Comment by DarthSim** [comment]
Ok, this is something. When Go falls with SIGSEGV, it prints lots of logs with stack trace and stuff. Try to grep your logs for `SIGSEGV`. If there is no mention of `SIGSEGV` in the imgproxy logs, this could mean that imgproxy is not a cause of the restarts.

**Comment by musabgultekin** [comment]
Its caused by alpine image. I've no clue why it happens but If you switch to Ubuntu image in docker, the issue will be gone. This is my imgproxy image based on Ubuntu and it works: [musabgultekin/imgproxy](https://github.com/musabgultekin/imgproxy)

**Comment by sneako** [comment]
Thanks @musabgultekin. I deployed the ubuntu image last night, and have not seen anymore restarts since then.

**Comment by DarthSim** [comment]
Oh, I finally got this. Musl which is used by Alpine has a very small stack size. libvips cache overflows the stack and causes SEGSEGV. I managed to reproduce the issue, and when I disabled the cache, the issue has gone. Well, we have two options here: disable cache or build docker image on top of Debian/Ubuntu.

**Comment by DarthSim** [comment]
This could tragically reduce the performance of libvips. Better kindly ask @jcupitt about this.

**Comment by jcupitt** [comment]
Hello, yes, musl (and therefore Alpine) has an 80kB stack, glibc (and therefore Ubuntu and indeed most linuxes) has at least 2MB. Unfortunately, musl provides no way to request a larger stack, as far as I know, and the developers are opposed to stack size increases, arguing that applications which need a larger stack are poorly designed.

We've tried to keep libvips stack use down, but it's difficult to fit a complex library within a stack that small. The libvips test suite will run correctly on Alpine but there are probably paths that are not tested. If you can find a simple example that overflows the musl stack, I can try to fix it.

The libvips cache helps with repeated operations. Suppose you have a pipeline like:

```
     A -> B \
             +-> D
A -> B -> C /
```

(sorry about the ascii art)

ie. the same sequence of images is repeated, then the cache will automatically transform it into:

```
A -> B -> C \
      \      +-> D
       -----/
```

So the `A -> B` is only computed once. This is very useful for complex applications where you don't know the pipeline in advance, but less useful for things like image resize services where the pipeline is fixed and can be optimised by a programmer. 

tldr: disabling the cache will probably not affect performance much, as long as you have tuned your pipeline.

**Comment by DarthSim** [comment]
@jcupitt Great! Thanks for the explanation!

I have tuned pipeline as much as I could (with your help, I still owe you a beer for this :) ). Since the pipeline is linear, looks like I can disable cache without any consequences.

Thanks a lot again, and good luck with libvips!

**Comment by DarthSim** [comment]
Released v1.1.8 with disabled libvips cache. Thanks, everyone!

**Comment by sneako** [comment]
I have deployed v1.1.8 and have not seen any restarts (or any noticeable performance degradation). Thanks everyone!
